shards/core/foundation.hpp

```diff
@@ -1536,7 +1534,12 @@ template <typename T> T &varAsObjectChecked(const SHVar &var, const shards::Type
     throw std::runtime_error(fmt::format("Invalid object type id, expected: {} got: {}", type,
                                          Type::Object(var.payload.objectVendorId, var.payload.objectTypeId)));
   }
-  return *reinterpret_cast<T *>(var.payload.objectValue);
+  if ((var.flags & SHVAR_FLAGS_CPP_SHARED_VOID_OBJECT) == SHVAR_FLAGS_CPP_SHARED_VOID_OBJECT) {
+    auto void_sp = static_cast<std::shared_ptr<void> *>(var.payload.objectValue);
+    return *std::static_pointer_cast<T>(*void_sp);
```

Comment:
This is wrong, we should detect that T could also be a shared ptr and extract only the inner type.